File: codesandbox/pkg/firecracker/network.go
package firecracker

import (
	"bufio"
	"bytes"
	"fmt"
	"os/exec"
	"regexp"
	"strings"
	"sync"
	"time"
)

// NetworkManager handles network configuration for VMs
type NetworkManager struct {
	bridgeName string
	subnet     string
}

// NewNetworkManager creates a new network manager
func NewNetworkManager(subnet string) *NetworkManager {
	return &NetworkManager{
		bridgeName: "fcbridge",
		subnet:     subnet,
	}
}

// Serialize TAP creation to avoid name races
var tapMu sync.Mutex

// Setup verifies the bridge exists and is up
func (nm *NetworkManager) Setup() error {
	out, err := exec.Command("ip", "link", "show", nm.bridgeName).CombinedOutput()
	if err != nil {
		return fmt.Errorf("bridge %s not found. Please run scripts/setup-host-network.sh first: %s", nm.bridgeName, strings.TrimSpace(string(out)))
	}
	if !strings.Contains(string(out), "UP") {
		if err := execCommand("sudo", "ip", "link", "set", nm.bridgeName, "up"); err != nil {
			return fmt.Errorf("failed to bring up bridge: %w", err)
		}
	}
	return nil
}

// Cleanup no-op; TAPs are per-VM
func (nm *NetworkManager) Cleanup() error { return nil }

// CreateTAPDeviceAuto creates a TAP with the next available name (fc-tap-N) and returns that name.
func (nm *NetworkManager) CreateTAPDeviceAuto() (string, error) {
	tapMu.Lock()
	defer tapMu.Unlock()

	// Try a small number of candidates to be resilient to races with other processes
	const maxCandidates = 64
	for i := 0; i < maxCandidates; i++ {
		name, err := nextFreeTapName("fc-tap")
		if err != nil {
			return "", err
		}
		// Attempt creation; if busy, loop to a new candidate
		if err := nm.createTapInternal(name); err != nil {
			if strings.Contains(err.Error(), "Device or resource busy") {
				// Someone beat us to it. Try another name after a tiny backoff.
				time.Sleep(50 * time.Millisecond)
				continue
			}
			return "", err
		}
		return name, nil
	}
	return "", fmt.Errorf("exhausted candidates while creating TAP device (busy)")
}

// CreateTAPDevice creates a TAP device with the requested name (idempotent-ish with cleanup & retry).
func (nm *NetworkManager) CreateTAPDevice(name string) error {
	tapMu.Lock()
	defer tapMu.Unlock()
	return nm.createTapInternal(name)
}

// Internal: create + attach + up + promisc with one cleanup+retry on busy.
func (nm *NetworkManager) createTapInternal(name string) error {
	// If device exists, try to delete it first
	if tapExists(name) {
		_ = safeDeleteTap(name)
		time.Sleep(150 * time.Millisecond)
	}

	err := execCommand("sudo", "ip", "tuntap", "add", "dev", name, "mode", "tap", "user", "dlo", "group", "dlo")
	if err != nil && strings.Contains(err.Error(), "Device or resource busy") {
		// Last-ditch cleanup + one retry—some processes keep fds briefly
		_ = safeDeleteTap(name)
		time.Sleep(200 * time.Millisecond)
		err = execCommand("sudo", "ip", "tuntap", "add", "dev", name, "mode", "tap", "user", "dlo", "group", "dlo")
	}
	if err != nil {
		return fmt.Errorf("failed to create TAP device: %w", err)
	}

	// Attach to bridge
	if err := execCommand("sudo", "ip", "link", "set", name, "master", nm.bridgeName); err != nil {
		_ = safeDeleteTap(name)
		return fmt.Errorf("failed to connect TAP to bridge: %w", err)
	}

	// Bring up + promisc
	if err := execCommand("sudo", "ip", "link", "set", name, "up"); err != nil {
		return fmt.Errorf("failed to bring up TAP device: %w", err)
	}
	if err := execCommand("sudo", "ip", "link", "set", name, "promisc", "on"); err != nil {
		return fmt.Errorf("failed to set TAP device to promiscuous mode: %w", err)
	}

	return nil
}

// DeleteTAPDevice removes a TAP device
func (nm *NetworkManager) DeleteTAPDevice(name string) error {
	return safeDeleteTap(name)
}

// ------- helpers -------

func tapExists(name string) bool {
	if err := exec.Command("ip", "link", "show", name).Run(); err == nil {
		return true
	}
	// Fallback check via `ip tuntap list`
	taps, _ := listTaps()
	for _, t := range taps {
		if t == name {
			return true
		}
	}
	return false
}

func safeDeleteTap(name string) error {
	_ = execCommand("sudo", "ip", "link", "set", name, "down")
	if err := execCommand("sudo", "ip", "link", "delete", name); err == nil {
		return nil
	}
	return execCommand("sudo", "ip", "tuntap", "del", "dev", name, "mode", "tap")
}

func listTaps() ([]string, error) {
	out, err := exec.Command("ip", "tuntap", "list").CombinedOutput()
	if err != nil {
		// Some distros may not support list; return empty set
		return nil, nil
	}
	var res []string
	sc := bufio.NewScanner(bytes.NewReader(out))
	re := regexp.MustCompile(`^([^\s:]+):`)
	for sc.Scan() {
		line := strings.TrimSpace(sc.Text())
		if m := re.FindStringSubmatch(line); len(m) == 2 {
			res = append(res, m[1])
		}
	}
	return res, nil
}

// nextFreeTapName finds the lowest fc-tap-N that doesn't exist (checks link & tuntap list)
func nextFreeTapName(prefix string) (string, error) {
	existing := map[string]struct{}{}

	// ip -o link shows all netdevs
	linkOut, _ := exec.Command("ip", "-o", "link").CombinedOutput()
	sc1 := bufio.NewScanner(bytes.NewReader(linkOut))
	re := regexp.MustCompile(`^\d+:\s*([^:@]+)`)
	for sc1.Scan() {
		line := sc1.Text()
		if m := re.FindStringSubmatch(line); len(m) == 2 {
			existing[m[1]] = struct{}{}
		}
	}

	// also add tuntap list
	if taps, _ := listTaps(); len(taps) > 0 {
		for _, t := range taps {
			existing[t] = struct{}{}
		}
	}

	// search a reasonable window
	const max = 4096
	for i := 0; i < max; i++ {
		name := fmt.Sprintf("%s-%d", prefix, i)
		if _, ok := existing[name]; !ok {
			// double-check with ip link show in case of TOCTOU
			if err := exec.Command("ip", "link", "show", name).Run(); err != nil {
				return name, nil
			}
		}
	}
	return "", fmt.Errorf("no free TAP names available (checked %d candidates)", max)
}

func execCommand(name string, args ...string) error {
	var cmd *exec.Cmd
	if name == "sudo" {
		cmd = exec.Command(name, args...)
		cmd.Env = append(cmd.Env, "SUDO_ASKPASS=/data/code/agentic-design/codesandbox/askpass.sh")
		newArgs := append([]string{"-A"}, args...)
		cmd.Args = append([]string{name}, newArgs...)
	} else {
		cmd = exec.Command(name, args...)
	}
	output, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("command '%s %s' failed: %v - output: %s", name, strings.Join(args, " "), err, strings.TrimSpace(string(output)))
	}
	return nil
}

File: codesandbox-rust/src/network.rs
use crate::{Result, CodeboxError};
use std::process::Command;
use std::sync::Mutex;
use std::thread::sleep;
use std::time::Duration;
use tracing::{debug, error};

pub struct NetworkManager {
    bridge_name: String,
    subnet: String,
}

// Serialize TAP creation to avoid name races
static TAP_MUTEX: Mutex<()> = Mutex::new(());

impl NetworkManager {
    pub fn new(subnet: String) -> Self {
        Self {
            bridge_name: "fcbridge".to_string(),
            subnet,
        }
    }

    pub fn setup(&self) -> Result<()> {
        let output = Command::new("ip")
            .args(&["link", "show", &self.bridge_name])
            .output()?;

        if !output.status.success() {
            return Err(CodeboxError::Network(format!(
                "bridge {} not found. Please run scripts/setup-host-network.sh first: {}",
                self.bridge_name,
                String::from_utf8_lossy(&output.stderr)
            )));
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        if !stdout.contains("UP") {
            self.exec_command(&["sudo", "ip", "link", "set", &self.bridge_name, "up"])?;
        }

        Ok(())
    }

    pub fn cleanup(&self) -> Result<()> {
        // No-op; TAPs are per-VM
        Ok(())
    }

    pub fn create_tap_device_auto(&self) -> Result<String> {
        let _guard = TAP_MUTEX.lock().unwrap();

        // Try a small number of candidates to be resilient to races with other processes
        const MAX_CANDIDATES: usize = 64;
        for _ in 0..MAX_CANDIDATES {
            let name = self.next_free_tap_name("fc-tap")?;
            
            // Attempt creation; if busy, loop to a new candidate
            match self.create_tap_internal(&name) {
                Ok(_) => return Ok(name),
                Err(e) if e.to_string().contains("Device or resource busy") => {
                    // Someone beat us to it. Try another name after a tiny backoff.
                    sleep(Duration::from_millis(50));
                    continue;
                }
                Err(e) => return Err(e),
            }
        }
        
        Err(CodeboxError::Network("exhausted candidates while creating TAP device (busy)".to_string()))
    }

    pub fn create_tap_device(&self, name: &str) -> Result<()> {
        let _guard = TAP_MUTEX.lock().unwrap();
        self.create_tap_internal(name)
    }

    fn create_tap_internal(&self, name: &str) -> Result<()> {
        // If device exists, try to delete it first
        if self.tap_exists(name) {
            let _ = self.safe_delete_tap(name);
            sleep(Duration::from_millis(150));
        }

        let mut err = self.exec_command(&[
            "sudo", "ip", "tuntap", "add", "dev", name, "mode", "tap", "user", "dlo", "group", "dlo"
        ]);

        if let Err(e) = &err {
            if e.to_string().contains("Device or resource busy") {
                // Last-ditch cleanup + one retry—some processes keep fds briefly
                let _ = self.safe_delete_tap(name);
                sleep(Duration::from_millis(200));
                err = self.exec_command(&[
                    "sudo", "ip", "tuntrap", "add", "dev", name, "mode", "tap", "user", "dlo", "group", "dlo"
                ]);
            }
        }

        if let Err(e) = err {
            return Err(CodeboxError::Network(format!("failed to create TAP device: {}", e)));
        }

        // Attach to bridge
        if let Err(e) = self.exec_command(&["sudo", "ip", "link", "set", name, "master", &self.bridge_name]) {
            let _ = self.safe_delete_tap(name);
            return Err(CodeboxError::Network(format!("failed to connect TAP to bridge: {}", e)));
        }

        // Bring up + promisc
        self.exec_command(&["sudo", "ip", "link", "set", name, "up"])
            .map_err(|e| CodeboxError::Network(format!("failed to bring up TAP device: {}", e)))?;
        
        self.exec_command(&["sudo", "ip", "link", "set", name, "promisc", "on"])
            .map_err(|e| CodeboxError::Network(format!("failed to set TAP device to promiscuous mode: {}", e)))?;

        Ok(())
    }

    pub fn delete_tap_device(&self, name: &str) -> Result<()> {
        self.safe_delete_tap(name)
    }

    fn tap_exists(&self, name: &str) -> bool {
        if Command::new("ip")
            .args(&["link", "show", name])
            .output()
            .map(|o| o.status.success())
            .unwrap_or(false)
        {
            return true;
        }

        // Fallback check via `ip tuntap list`
        if let Ok(taps) = self.list_taps() {
            return taps.iter().any(|t| t == name);
        }

        false
    }

    fn safe_delete_tap(&self, name: &str) -> Result<()> {
        let _ = self.exec_command(&["sudo", "ip", "link", "set", name, "down"]);
        
        if self.exec_command(&["sudo", "ip", "link", "delete", name]).is_ok() {
            return Ok(());
        }
        
        self.exec_command(&["sudo", "ip", "tuntap", "del", "dev", name, "mode", "tap"])
    }

    fn list_taps(&self) -> Result<Vec<String>> {
        let output = Command::new("ip")
            .args(&["tuntap", "list"])
            .output()?;

        if !output.status.success() {
            // Some distros may not support list; return empty set
            return Ok(vec![]);
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        let mut result = vec![];
        
        for line in stdout.lines() {
            let line = line.trim();
            if let Some(name) = line.split(':').next() {
                result.push(name.to_string());
            }
        }

        Ok(result)
    }

    fn next_free_tap_name(&self, prefix: &str) -> Result<String> {
        let mut existing = std::collections::HashSet::new();

        // ip -o link shows all netdevs
        if let Ok(output) = Command::new("ip").args(&["-o", "link"]).output() {
            let stdout = String::from_utf8_lossy(&output.stdout);
            for line in stdout.lines() {
                if let Some(captures) = regex::Regex::new(r"^\d+:\s*([^:@]+)")
                    .unwrap()
                    .captures(line)
                {
                    if let Some(name) = captures.get(1) {
                        existing.insert(name.as_str().to_string());
                    }
                }
            }
        }

        // also add tuntap list
        if let Ok(taps) = self.list_taps() {
            for tap in taps {
                existing.insert(tap);
            }
        }

        // search a reasonable window
        const MAX: usize = 4096;
        for i in 0..MAX {
            let name = format!("{}-{}", prefix, i);
            if !existing.contains(&name) {
                // double-check with ip link show in case of TOCTOU
                if !Command::new("ip")
                    .args(&["link", "show", &name])
                    .output()
                    .map(|o| o.status.success())
                    .unwrap_or(false)
                {
                    return Ok(name);
                }
            }
        }

        Err(CodeboxError::Network(format!(
            "no free TAP names available (checked {} candidates)", MAX
        )))
    }

    fn exec_command(&self, args: &[&str]) -> Result<()> {
        debug!("Executing command: {:?}", args);
        
        let mut cmd = if args[0] == "sudo" {
            let mut cmd = Command::new("sudo");
            cmd.env("SUDO_ASKPASS", "/data/code/agentic-design/codesandbox/askpass.sh");
            cmd.arg("-A");
            cmd.args(&args[1..]);
            cmd
        } else {
            let mut cmd = Command::new(args[0]);
            cmd.args(&args[1..]);
            cmd
        };

        let output = cmd.output()?;
        
        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            error!("Command failed: {:?}, output: {}", args, stderr);
            return Err(CodeboxError::Network(format!(
                "command '{:?}' failed: {}",
                args, stderr
            )));
        }

        Ok(())
    }
}